File: problems/to-categorize/Math and Geometry/13: Roman To Integer.js
// https://leetcode.com/problems/roman-to-integer/description/
// Difficulty: Easy
// tags: arrays

// Solution, O(1) time and O(1) space. O(1) time since the number of valid roman numerals with the given characters is bounded.
/*
Iterate over the roman numeral string. If the current letter is less than the next number, like IV, add the difference and increment past the second letter. Otherwise just add the value as normal.
*/

const VALUES = {
  I: 1,
  V: 5,
  X: 10,
  L: 50,
  C: 100,
  D: 500,
  M: 1000,
};

var romanToInt = function (s) {
  let result = 0;
  let pointer = 0;
  while (pointer < s.length) {
    const letter = s[pointer];
    const value = VALUES[letter];
    const letterNext = s[pointer + 1];
    const valueNext = VALUES[letterNext];
    if (value < valueNext) {
      result += valueNext - value;
      pointer += 2;
    } else {
      result += value;
      pointer++;
    }
  }
  return result;
};

File: problems/Leetcode/13: Roman To Integer.py
VALS = { 'I' : 1, 'V' : 5, 'X' : 10, 'L' : 50, 'C' : 100, 'D' : 500, 'M' : 1000 }

class Solution:
    def romanToInt(self, s: str) -> int:
        result = 0
        i = 0
        while i < len(s):
            if i == len(s) - 1:
                return result + VALS[s[i]]
            # if the next value is bigger than the current one, we add the difference
            next_val = VALS[s[i + 1]]
            curr_val = VALS[s[i]]
            if next_val > curr_val:
                result += next_val - curr_val
                i += 1
            else:
                result += curr_val
            i += 1        
        return result